apps/libsignpost/port_signpost_arduino.cpp

```diff
@@ -0,0 +1,252 @@
+//Slave receive buffer
+static uint8_t* g_slave_receive_buf = NULL;
+static size_t g_slave_receive_buf_max_len = 0;
+static size_t g_slave_receive_buf_len = 0;
+//Slave transmission buffer
+static uint8_t* g_slave_transmit_buf = NULL;
+static size_t g_slave_transmit_buf_len = 0;
+
+int port_signpost_init(uint8_t i2c_address) {
+	//Initialize i2c as slave
+	Wire.begin(i2c_address);
+	g_arduino_i2c_address = i2c_address;
+	//Setup GPIO pins
+	pinMode(ARDUINO_MOD_IN, INPUT);
+	pinMode(ARDUINO_MOD_OUT, OUTPUT);
+	pinMode(ARDUINO_PPS, INPUT);
+	pinMode(DEBUG_LED, OUTPUT);
+	//Set callback helpers for receiving and transmitting i2c message in slave mode
+	Wire.onReceive(slave_listen_callback_helper);
+	Wire.onRequest(slave_read_callback);
+	return 0;
+}
+
+int port_signpost_i2c_master_write(uint8_t addr, uint8_t* buf, size_t len) {
+	if (len == 0) {
+		return -1;
+	}
+	//Initialize i2c as master
+	Wire.begin();
+	//Set i2c clock speed
+	Wire.setClock(SIGNPOST_I2C_SPEED);
+	//Send i2c message to addr
+	Wire.beginTransmission(addr);
+	int num_written = Wire.write(buf, len);
+	int rc = Wire.endTransmission();
+	//Put Arduino back into i2c slave mode
+	Wire.begin(g_arduino_i2c_address);
+	return num_written;
+}
+
+int port_signpost_i2c_slave_listen(port_signpost_callback cb, uint8_t* buf, size_t max_len) {
+	//Initialize i2c as slave
+	Wire.begin(g_arduino_i2c_address);
+	//Set i2c slave listen callback and buffer
+	g_slave_listen_callback = cb;
+	g_slave_receive_buf = buf;
+	g_slave_receive_buf_max_len = max_len;
+	g_slave_receive_buf_len = 0;
+	return 0;
+}
+
+//Function untested
+int port_signpost_i2c_slave_read_setup(uint8_t* buf, size_t len) {
+	//Initialize i2c as slave
+	Wire.begin(g_arduino_i2c_address);
+	//Set global i2c slave transmit buffer
+	g_slave_transmit_buf = buf;
+	g_slave_transmit_buf_len = len;
+	return 0;
+}
+
+int port_signpost_mod_out_set() {
+	digitalWrite(ARDUINO_MOD_OUT, HIGH);
+	return 0;
+}
+
+int port_signpost_mod_out_clear() {
+	digitalWrite(ARDUINO_MOD_OUT, LOW);
+	return 0;
+}
+
+int port_signpost_mod_in_read() {
+	return digitalRead(ARDUINO_MOD_IN);
+}
+
+int port_signpost_mod_in_enable_interrupt_falling(port_signpost_callback cb) {
+	//Set global MOD_IN falling edge callback 
+	g_mod_in_callback_falling = cb;
+	//Attach callback helper to hardware interrupt
+	attachInterrupt(digitalPinToInterrupt(ARDUINO_MOD_IN), mod_in_callback_falling_helper, FALLING);
+	return 0;
+}
+
+int port_signpost_mod_in_enable_interrupt_rising(port_signpost_callback cb) {
+	//Set global MOD_IN rising edge callback 
+	g_mod_in_callback_rising = cb;
+	//Attach callback helper to hardware interrupt
+	attachInterrupt(digitalPinToInterrupt(ARDUINO_MOD_IN), mod_in_callback_rising_helper, RISING);
+	return 0;
+}
+
+int port_signpost_mod_in_disable_interrupt(void) {
+	detachInterrupt(digitalPinToInterrupt(ARDUINO_MOD_IN));
+	return 0;
+}
+
+//TODO: Implement version that puts Arduino into a sleep state
+void port_signpost_wait_for(void* wait_on_true) {
```

Comment:
It's cast into a volatile within that function, but that should probably changed to volatile in the port layer. I was told to implement it this way for the moment.